lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -960,12 +978,67 @@ void Inliner::inlineInstances(FModuleOp parent) {
   }
 }
 
+void Inliner::identifyNLAsTargetingOnlyModules() {
+  DenseSet<FModuleOp> nlaTargetedModules;
+
+  // Identify candidate NLA's: those that end in a module
+  for (auto &[sym, mnla] : nlaMap) {
+    auto nla = mnla.getNLA();
+    if (nla.isModule())
+      nlaTargetedModules.insert(symbolTable.lookup<FModuleOp>(nla.leafMod()));
+  }
+
+  // Helper to scan leaf modules for users of NLAs, gathering by symbol names
+  auto scanForNLARefs = [&](FModuleOp mod) {
+    DenseSet<StringAttr> referencedNLASyms;
+    // Scan ports
+    for (unsigned i = 0, e = mod.getNumPorts(); i != e; ++i)
+      for (auto anno : AnnotationSet::forPort(mod, i))
+        if (auto sym = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal"))
+          referencedNLASyms.insert(sym.getAttr());
+
+    // Scan operations (and not the module itself):
+    mod.getBody()->walk([&](Operation *op) {
+      for (auto anno : AnnotationSet(op))
+        if (auto sym = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal"))
+          referencedNLASyms.insert(sym.getAttr());
+    });
+
+    return referencedNLASyms;
+  };
```

Comment:
@youngar has brought up on a PR of mine that this isn't checking for two cases (which may or may not matter): (1) annotations on instance ports and (2) annotations on memory ports.

While I currently think a large part of this is unreachable (e.g., https://github.com/llvm/circt/issues/3340 shows the inability of CIRCT to even generate things of type (1) above), it would still be good to iterate over the ports of memories and instances to dig out any annotations that may reference hierpath symbols.

Reply:
Since the ModuleInliner itself doesn't look into these types of annotations, I'm not sure what happens if these are encountered.  I added support for MemOp ports to this portion of the code, and will look into InstanceOp port annotations-- will add that shortly-- `AnnotationSet` doesn't have the nice helpers for it, and FWIW `PortAnnoTarget::getType()` checks for `InstanceOp` (in same place checks for `MemOp`) but there's no constructor that makes that possible AFAICT.  Anyway, will add that shortly.  Better safe than sorry re:removing.

This all makes it seem, as mentioned previously elsewhere, that having something cleanup dead HierPathOp's would be a good idea, and would largely use this sort of code (or similar from the other NLA migration PR's), where we definitely want to handle all these things (regardless of what's reachable here :smile:).

Reply:
😂  I didn't realize that `ModuleInliner` wasn't even looking at these. That is some empirical evidence that this may be pretty rare. Thanks for adding this, though.